### DISCUSSION TITLE
Fatal error while creating a multi-user code-collab platform using code-server.

### DISCUSSION THREAD

**Discussion by masterMav** [coder/code-server#6311] [Q&A]
My goal is to create a multi-file collaborative code-editor where multiple people from anywhere on the internet can join and edit the same file or project simultaneously with real-time updates in the browser itself.

**My approach**
After installing code-server ([using this](https://coder.com/docs/code-server/latest/install#debian-ubuntu)) for my laptop running ubuntu 22.04.2 LTS, I found out that if I run code-server to open vscode in the browser on port 8080. I can then copy the same url (i.e http://127.0.0.1:8080/) and paste it in some other browser like firefox or open in incognito, then I have these two tabs running vscode in the browser. After changing the autosave settings to afterdelay with a 10 ms gap, I can make changes in any file and those changes are reflected in other tabs as well in real-time. So my requirement of creating a collaborative code-editor is fulfilled.

Now, I want to do the same by hosting a code-server in my ec2 instance where multiple people from the internet can join and code-collab in the browser itself since any change made to any file is reflected on other tabs in real-time using auto-save. However, this is not working like it did in my local machine.

**Problems faced**
1. code-server crashes when two or more tabs are opened with the same url. When one tab is opened it works fine, but when another tab is opened it disconnects and asks for reload until only one tab is open, nothing more.

2. When I used ngrok to open the same localhost port on my friends laptop, any change I make in any file is reflected to him as well but as soon as we start making change to the same file at the same time, one of us gets an error message stating file changes not saved compare or overwrite? So basically we can only make changes one at a time but that change is shown on every connected machine instantly.

On AWS, I’m running an ec2 instance (ubuntu platform) where I have installed code-server ([using this](https://coder.com/docs/code-server/latest/install#debian-ubuntu)) and nginx setup ([using this](https://github.com/coder/code-server/blob/v3.8.1/docs/guide.md#nginx)). 

Any help or suggestions will be greatly appreciated.

**Comment by lg-nineoneninethreenine** [comment]
I believe that code-server on your EC2 instance is first saving the files as localStorage and then uploading it. Now if your friend were to edit the same file you are editing, the files would have two versions which don't coincide with each other, hence the request to overwrite. You could rather use [Duckly](https://duckly.com) and Git for version control too. That way, pair programming with version control is available and can be used.

**Reply by lg-nineoneninethreenine** [reply]
Never use the same code-server instance in two different places at the same time except if it is you, since doing so would create two versions of the same file that don't coincide enough to be saved, rather they split into copies.

**Comment by code-asher** [comment]
Yeah code-server has no support for merging concurrent edits to the same file so whoever saved first would end up "winning" and the other person would have to either overwrite or manually compare the changes.  Like @im-coder-lg mentioned a third-party tool like Duckly could work.

As for the disconnects, we will need to see logs to figure those out.  Multiple tabs opened to the same URL should work, although this is not an explicitly supported use case.

There is an open issue for multi-user support (https://github.com/coder/code-server/issues/33) but I do not see it being implemented any time soon and I am not sure code-server is the right place to implement it anyway.
